[BCFR-899] MaxLogsKept implementation

reductionista opened the pull request:
[BCFR-899](https://smartcontract-it.atlassian.net/browse/BCFR-899)

# Motivation
Presently, LogPoller supports only time-based retention, via the Retention field in the filters passed to RegisterFilter. The MaxLogsKept field was added earlier in anticipation of the need for also supporting recency-count based retention. One example of a case where time based retention is risky is the Transmit event in the OCR Contract Transmitter. No matter how long the retention period is set to, there's a chance the node will be down for longer than that, and miss logs when it comes back up. This would make a bad situation even worse, because the transmit event would never be picked up at all.

# Solution
This implements the MaxLogsKept feature in LogPoller. When specified, this field tells LogPoller it's okay to prune logs matching a filter if there are at least MaxLogsKept more recent matching logs in the db.
In the example above, this avoid storing any more logs than needed while always having the latest transmit event available. In this case, older transmit events are no longer relevant if there is a more recent one.
In general, this should be just as useful for anything accessed only via ChainReader's GetLatestValue() method rather than QueryKey().

A log may be pruned either because it's too old in terms of time or in terms of the number of logs being saved. It need not satisfy both theRetention and MaxLogsKeptcriteria in order to get pruned.


[BCF-899]: https://smartcontract-it.atlassian.net/browse/BCF-899?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[BCFR-899]: https://smartcontract-it.atlassian.net/browse/BCFR-899?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

# Testing

This was tested by changing the MaxLogsKept setting on the ContractTransmitter filter passed to LogPoller from 0 to 1, and running the CCIP load tests.

Without paging, this query is one of the most cpu intensive. Similar to the DeleteExpiredLogs query, it must go though every row of the logs table unless LogPrunePageSize is set to non-zero.  But it's slower than DeleteExpiredLogs because on top of that, it also has to group, sort, and count every log in each group in order to figure out how many there are and which ones are the excess logs eligible for deletion. Also similar to DeleteExpiredLogs, the final step is to merge together all the results for logs matching multiple filters by making sure that no log is dropped unless ALL of its matching filters agree that it came before their own MaxLogsKept threshold was hit.

Without paging, the SelectExcessLogs query grew in median duration linearly as the number of logs in the table increased. After a few hours it was taking longer than the Insert queries (and any other queries) and started causing timeouts to occur, not just for itself but for other queries as well. It eventually got to a point where the p90 & p99 charts were continuously in the 4-5s range, generating many critical errors as well as a lot of backlogged queries waiting on connections.

With a paging size of 4000, you could see that the query durations grew linearly at first with the size of the table, and then leveled off at slightly more than the insert queries. It still resulted in a couple fairly large bursts of critical errors (query timeouts) during the heaviest cpu usage.

A paging size of 1000-2000 worked much better. There were a much smaller # of timeouts, and only during a brief window of time.  The p99 durations were noticeably lower than the insert query durations, aside from some very brief but high spikes which we believe are due to an unrelated bug (which has been fixed in Chainlink repo, but hasn't been back posted to CCIP yet) in not letting go of connections quickly enough (due to sql logging).
Aside from these spikes, the charts look pretty healthy... so we should retest once that bug is backported but before the MaxLogsKept feature is enabled.

Another scenario tested was the performance of the query when all filters have MaxLogsKept=0. This was tested both with and without the code which skips the query unless there is at least one filter with MaxLogsKept > 0. Obviously, skipping the query took it off the chart entirely... so the rest of the chart looked the same as before this PR. Without slipping the query, it was slightly faster than the MaxLogs=1 case, but not a very significant reduction. Based on this it was decided it did make sense to disable the pruning until this feature starts being used by at least 1 filter.

With MaxLogsKept=1:
<img width="2496" alt="Screenshot 2024-10-15 at 5 54 53 PM" src="https://github.com/user-attachments/assets/fb97be8d-2e67-432b-80c0-54dcdb2d8e86">
<img width="2487" alt="Screenshot 2024-10-15 at 5 55 11 PM" src="https://github.com/user-attachments/assets/4e179f97-4e81-422a-bece-89eed1be6246">

With MaxLogsKept=0 (disabled):
<img width="2489" alt="Screenshot 2024-10-15 at 5 58 09 PM" src="https://github.com/user-attachments/assets/f8de0a00-356e-4b98-a0f9-fb0eb9df735c">
<img width="2516" alt="Screenshot 2024-10-15 at 5 58 46 PM" src="https://github.com/user-attachments/assets/17109b1a-96f6-4c3e-979e-26c745acac28">

For both of these tests, the pruning of unmatched logs--which is also an expensive query--has been increased in frequency from every 20 ticks to every 4 ticks, to see what it looks like under the most severe conditions while they're both happening at once (with every 20 ticks, the test usually completes before the first one happens).  On the MaxLogsKept=1 charts, SelectUnmatchedLogs query durations show up as cyan and SelectExcessLogs query durations as bright yellow. On the MaxLogsKept=0 charts, SelectUnmatchedLogs is pale yellow (and SelectExcessLogs does not run--as desired). LogPrunePageSize was set to 1000 for both.

reductionista committed:
Add PruneExcessLogs for MaxLogsKept

reductionista committed:
Update contract_transmitter.go for MaxLogsKept

reductionista committed:
Fix unreachable code

reductionista committed:
pnpm changeset

dhaidashenko commented on core/chains/evm/logpoller/orm.go:
```diff
@@ -400,6 +401,45 @@ func (o *DSORM) SelectUnmatchedLogIDs(ctx context.Context, limit int64) (ids []u
+	}
+
+	// Allow SELECT query to run for up to 3 minutes. DELETE query will still have default 10s timeout
+	selectCtx, cancel := context.WithTimeout(sqlutil.WithoutDefaultTimeout(ctx), 3*time.Minute)
```
It might be a good idea to check if we have a filter with `max_logs_kept > 0` before running the main query

dhaidashenko commented on core/chains/evm/logpoller/orm.go:
```diff
@@ -400,6 +401,45 @@ func (o *DSORM) SelectUnmatchedLogIDs(ctx context.Context, limit int64) (ids []u
 }
 
+// SelectExcessLogIDs finds any logs old enough that MaxLogsKept has been exceeded for every filter they match.
+func (o *DSORM) SelectExcessLogIDs(ctx context.Context, limit int64) (rowIDs []uint64, err error) {
```
Should we have a test for it?

dhaidashenko commented on core/chains/evm/logpoller/log_poller.go:
```diff
@@ -1121,13 +1125,35 @@ func (lp *logPoller) PruneOldBlocks(ctx context.Context) (bool, error) {
+		done = false
+	}
+
+	rowIDs, err := lp.orm.SelectExcessLogIDs(ctx, lp.logPrunePageSize)
```
Should we have a feature flag to disable it if it causes issues in production?

reductionista replied:
I took your other suggestion of not calling `PruneExcessLogs()` unless there is a filter registered with `MaxLogsKept > 0`.  That may be good enough for now, since there aren't even any products yet which set `MaxLogsKept > 0`

When this gets merged into the ccip repo and they are ready to start passing non-zero values for it, they can decide whether to make them configurable, hide them behind a feature flag, or just always hard code them to non-zero.  If it's the last they opt for, we can revisit adding a global feature flag for it, but having things like Retention and MaxLogsKept be configurable per product seems like something we ought to encourage, rather than an all-or-nothing approach.  For example, they could have a feature flag which sets either the Retention field or MaxLogsKept... but having both at once seems unnecessary and having only a MaxLogsKept that can be turned on or off in LogPoller's global configuration would mean turning it off goes back to "retain everything" rather than a time-based retention.  Curious to hear what @mateusz-sekara thinks though.

mateusz-sekara replied:
I think enabling/disabling that with a feature flag adds another layer of complexity to understand of what is enabled and what not (and would probably make code complex as well). I would refrain from using it unless it's necessary. As @reductionista mentioned, this feature is not used by anyone else and probably we will be the first customer of that. This was already tested pretty heavily by @reductionista, but we will let it soak on pre-prod environment for some time before going to mainnet. That should give us a lot insights on what's the impact and how the query behaves in the long run.

In case of issues on prod-mainnet we can disable that by changing how the filter is configured instead of introducing another feature flag (bring back MaxLogsKept to the previous value). Does it make sense?

dhaidashenko replied:
Sounds good

dhaidashenko reviewed (commented):
Impressive query 🤯!
Left a couple of nits

reductionista committed:
Re-use block-range paging from DeleteBlocksBefore for SelectExcessLogs

Also: add block_number >= lower

reductionista committed:
Use ExecPagedQuery for SelectUnmatchedLogIDs

Also: add deduping of ids for logs matching multiple filters

reductionista committed:
Improve logging

reductionista committed:
Add test for SelectExcessLogIDs

Also, remove some extraneous lines in orm_test.go

reductionista committed:
Only activate count-based log pruning when needed

reductionista commented on core/chains/evm/logpoller/log_poller.go:
```diff
@@ -217,6 +218,12 @@ func (filter *Filter) Contains(other *Filter) bool {
+	}
+	if other.MaxLogsKept != filter.MaxLogsKept {
+		return false
+	}
```
I don't think there are any current products which actually use the "update a filter" feature (same filter name with different addresses or events), but for consistency we should have added the `other.Retention != filter.Retention` check when support for `Retention` field was added. Adding this now so that `Retention` and `MaxlogsKept` can be updated dynamically if needed.

mateusz-sekara commented on core/chains/evm/logpoller/log_poller.go:
```diff
@@ -665,31 +675,41 @@ func (lp *logPoller) backgroundWorkerRun() {
 				// Tick faster when cleanup can't keep up with the pace of new logs
 				logPruneTick = tickWithDefaultJitter(logPruneShortInterval)
-			} else if successfulExpiredLogPrunes == 20 {
+			} else if successfulExpiredLogPrunes >= 20 {
```
👍

mateusz-sekara commented on core/chains/evm/logpoller/orm.go:
```diff
@@ -386,25 +393,96 @@ type Exp struct {
+		for _, id := range rowIDs {
+			if _, val := m[id]; !val {
+				m[id] = true
+				ids = append(ids, id)
```
This looks a bit hacky, doesn't it? From the lambda function we are modyfing slice from the outer scope. Shouldn't we make `ExecPagedQuery` more configurable to just return these ids?

mateusz-sekara replied:
I wonder if something like this would make it simpler in this case
```go
	_, resultIDs, err := o.ExecPagedQuery(ctx, limit, latestBlock.FinalizedBlockNumber, func(lower, upper int64) (int64, uint64, error) {
		var rowIDs []uint64
		err2 = o.ds.SelectContext(ctx, &rowIDs, query, ubig.New(o.chainID), lower, upper)
		return int64(len(rowIDs)), rowIDs, err2
	})
	
	uniqueIds := make(map[uint64]struct{}{}, len(resultIDs))
	for _, id := range resultIDs {
		uniqueIds[id] = struct{}{}
	}

	return maps.Keys(uniqueIds), err
```

```go
// acc is a simple accumulator, it's joined by the ExecPagedQuery function, meaning that 
// finalAcc := acc_1 + acc_2 + ... + acc_N 
//If generics are the problem it could be `acc []interface{}`
func ExecPagedQuery[T comparable](ctx context.Context, limit int64, end int64, query func(lower, upper int64) (int64, []T, error)) (numResults int64, acc []T, err error) {}
```

Then you should be able to use `ExecPagedQuery` with any query returning any results

dhaidashenko replied:
In which cases is it possible to get log ID duplicates?
Does `ExecPagedQuery` not go through distinct block ranges?

reductionista replied:
@dhaidashenko There will be duplicates whenever a log matches more than one filter, due to the JOIN with the filters table.

reductionista replied:
@dhaidashenko Actually, I guess you're right... they can't match with multiple filters in the Unmatched Logs query since they won't match with any filter, and the Excess Logs query has a GROUP BY id that should get rid of them. Must have gotten confused!

reductionista replied:
> // acc is a simple accumulator, it's joined by the ExecPagedQuery function, meaning that // finalAcc := acc_1 + > acc_2 + ... + acc_N //If generics are the problem it could be `acc []interface{}` func ExecPagedQuery[T > comparable](ctx context.Context, limit int64, end int64, query func(lower, upper int64) (int64, []T, error)) > (numResults int64, acc []T, err error) {}

I thought of using generics for this, but it unfortunately doesn't allow them on methods. I'll see if I can get it to work with []interface{}

reductionista replied:
I tried doing this with `[]interface{}` but it doesn't seem possible without some pretty heavy use of `reflect`, due to not being able to directly append to or cast a slice of an unknown type:
```
if reflect.TypeOf(rows).Kind() == reflect.Slice {
                    finalAcc = reflect.AppendSlice(reflect.ValueOf(finalAcc), reflect.ValueOf(rows))
                } else {
```
I decided that seems even more hacky than the closure. Maybe should have just gone back to my original at that point, but I ended up finding a way to do it with generics after all. It's also less than ideal, because I had to remove it from the `ORM` in order to use the generics which means you have to pass around the data source and chain id. Overall, I think it's an improvement though, because it seems less error prone and more flexible for adapting it to other purposes in the future.

mateusz-sekara commented on core/chains/evm/logpoller/orm.go:
```diff
@@ -386,25 +393,96 @@ type Exp struct {
+		var rowIDs []uint64
+		err2 = o.ds.SelectContext(ctx, &rowIDs, query, ubig.New(o.chainID), lower, upper)
+		// dedupe rowIDs before appending them to results
+		m := make(map[uint64]bool)
```
this map is in the scope of lambda function, so if you have duplicates between the first call and the Nth call (e.g. first page and second page) they won't be deduped in the final result set, right?

reductionista replied:
Logs from different pages will have different block numbers, so there can't be any duplicates there

mateusz-sekara commented on core/chains/evm/logpoller/orm.go:
```diff
@@ -386,25 +393,96 @@ type Exp struct {
+			FROM evm.log_poller_filters WHERE evm_chain_id=$1
+			GROUP BY name`
+
+	countLogsSubQuery := ` -- Count logs matching each filter in reverse order, labeling anything after the filter.max_logs_kept'th with old=true
```
How about having these comments in go code instead of passing them in the query as comments?

mateusz-sekara commented on core/chains/evm/logpoller/log_poller.go:
```diff
@@ -1121,13 +1142,39 @@ func (lp *logPoller) PruneOldBlocks(ctx context.Context) (bool, error) {
+		done = false
+	}
+
+	if !lp.countBasedLogPruningActive.Load() {
```
Why do we need this check? Are we concerned about `SelectExcessLogIDs` query performance when nothing is matched? Ideally when there are no filters with `MaxLogsKept>0` the `SelectExcessLogIDs` should be insanely fast.

If we have a flag for `countBasedLogPruningActive` why don't we have a flag for `timeRetentionBasedLogPruningActive`? ;)

reductionista replied:
> Are we concerned about SelectExcessLogIDs query performance when nothing is matched? Ideally when there are no filters with MaxLogsKept>0 the SelectExcessLogIDs should be insanely fast.

I don't know of any easy way of making the query any faster for the case where they're all zero, aside from just adding an extra query before it to check for that special case.  We can't just ignore rows where MaxLogKept=0 because each of them affects all other rows (if any log joins with both that one and another row with MaxLogsKept > 0 then we can't drop it). Only when they happen to be all 0's can any of them be ignored... which is the kind of thing the planner is just not designed to figure out.

So one way to think about the purpose of this `lp.countBasedLogPruneActive` flag is, it's just an optimization so that we don't have to add an extra query to check for that special case every time.

The same is true for the retention query, but that one is by nature a lot of faster because it doesn't have to sort and count subsets of logs... it only has to do a single comparison.  So it's less savings, plus it's already in use... so I'm not sure it would matter any more--the most cpu/memory savings would be now when the feature is first being introduced and nobody is using it yet. It also gives us something fairly similar to a "feature flag" where we know that it can be rolled out without hurting anything while it continues to get tested in new ways as products start to make use of it on develop branches.

I wanted to confirm all of the above by actually load testing it, but for the past week the load tests have been broken on `ccip-develop` branch (until today, I kept thinking it must have been my change, but apparently it was due to ChainReader getting enabled in the docker file). Tonight I finally got the load test to work again, by rebasing onto an earlier commit in ccip repo from a couple weeks ago. The test confirms that the query times for SelectExcessLogIDs are about the same (maybe slightly faster, but not a significant difference) with all MaxLogKept=0 rows as when some rows are non-zero.

So I've left it in there, and addressed @dhaidashenko 's comment about having not fully implemented it the previous time.

mateusz-sekara commented on core/chains/evm/logpoller/orm.go:
```diff
@@ -386,25 +393,96 @@ type Exp struct {
+	withSubQuery := ` -- Roll up the filter table into 1 row per filter
+		SELECT name,
+				ARRAY_AGG(address) AS addresses, ARRAY_AGG(event) AS events,
+				(ARRAY_AGG(topic2) FILTER(WHERE topic2 IS NOT NULL)) AS topic2,
```
Are these topics used for any filtering at this moment? Maybe we should not pollute the query by aggregating these topics if we don't need them, wdyt?

reductionista replied:
Not at the moment, but I've been trying to include them in any new queries we add so that it will be less work when we eventually merge the PR that implements them. I don't think they affect the query time noticeably.

reductionista replied:
Not at the moment, but I've been trying to include them in any new queries we add so that it will be less work when we eventually merge the PR that implements them. I don't think they affect the query time noticeably.

mateusz-sekara replied:
It's more of a readability thing. If I go to code and see these options used in the query I might incorrectly assume this feature is fully supported and should just work. If you plan to implement that soon we can keep it, but if we don't have any exact plans about supporting that feature then we can drop them. Kinda YAGNI rule. Up to you ;)

reductionista replied:
Alright

dhaidashenko commented on core/chains/evm/logpoller/log_poller.go:
```diff
@@ -282,14 +289,17 @@ func (lp *logPoller) RegisterFilter(ctx context.Context, filter Filter) error {
 	lp.filters[filter.Name] = filter
 	lp.filterDirty = true
+	if filter.MaxLogsKept > 0 {
+		lp.countBasedLogPruningActive.Store(true)
```
Is not it possible that a filter with `MaxLogsKept > 0` will be loaded from DB and we won't set countBasedLogPruningActive to true?

reductionista replied:
Yikes! Yes, a big oversight on my part. @mateusz-sekara just left a comment expressing skepticism about having this flag at all... so I guess I need to decide whether to just back out this change or add a check where `LoadFilters()` is called so that it always gets set if needed.

I'm kind of on the fence, I guess what I'll do is gather data on how fast the query is when `MaxLogsKept=0` for all filters and decide based on that.

dhaidashenko commented on core/chains/evm/logpoller/orm.go:
```diff
@@ -320,27 +317,37 @@ func (o *DSORM) DeleteBlocksBefore(ctx context.Context, end int64, limit int64)
-
-		if rows, err = result.RowsAffected(); err != nil {
-			return deleted, err
+		rows, err2 := query(lower, upper)
```
Here you return `err` that is nil instead of `err2`. Is it desired? 
Is it not cleaner and less error-prone to reuse err instead of defining a new err2 variable?

reductionista replied:
Good catch, thanks!

reductionista committed:
Typo in changeset

reductionista committed:
Refactor ExecPagedQuery into method of generic type RangedQuery[T]

reductionista commented:
As far as I can tell from the unit and integration tests, everything is passing. But since Thurs I've been having trouble getting the ccip load tests to run, so I still haven't been able to test the performance of the query when MaxLogsKept=0 for all filters.
At first I was getting this error:
```
    log.go:43: 00:41:51.16 INF UsdPerTokenUpdated event received Lane=private-chain-10-->private-chain-8 chain=private-chain-10 env=load-ccip-3c00f price_registry=0x5E388db7c3aef7deA9585FbE3c077f0c98857d0d token=0xeD1DB453C3156Ff3155a97AD217b3087D5Dc5f6E
    log.go:43: 00:41:51.16 INF UsdPerTokenUpdated event received Lane=private-chain-10-->private-chain-17 chain=private-chain-10 env=load-ccip-3c00f price_registry=0x5E388db7c3aef7deA9585FbE3c077f0c98857d0d token=0xeD1DB453C3156Ff3155a97AD217b3087D5Dc5f6E
2024/10/08 00:41:51.289506 WARN RESTY Post "http:/XX.XX.XX.XX:6688/v2/jobs": dial tcp XX.XX.XX.XX:6688: connect: connection refused, Attempt 4
2024/10/08 00:41:51.289534 ERROR RESTY Post "http://XX.XX.XX.XX:6688/v2/jobs": dial tcp XX.XX.XX.XX:6688: connect: connection refused
    ccip.go:1190:
        	Error Trace:	/go/testdir/integration-tests/ccip-tests/testsetups/ccip.go:1190
        	            				/go/testdir/integration-tests/ccip-tests/load/helper.go:94
        	            				/go/testdir/integration-tests/ccip-tests/load/ccip_test.go:87
        	Error:      	Received unexpected error:
        	            	shouldn't fail creating CCIP-Commit job on OCR node 6 job name ccip-commit-private-chain-6-private-chain-12 - Post "http://XX.XX.XX.XX:6688/v2/jobs": dial tcp XX.XX.XX.XX:6688: connect: connection refused
        	Test:       	TestLoadCCIPStableRequestTriggeringWithNetworkChaos
        	Messages:   	Creating jobs shouldn't fail

```
But after rebasing everything, I'm getting this error now (about 25 mins after the remote runner starts):
```
        	Error Trace:	/go/testdir/integration-tests/ccip-tests/testsetups/ccip.go:931
        	            				/go/testdir/integration-tests/ccip-tests/testsetups/ccip.go:1192
        	            				/go/testdir/integration-tests/ccip-tests/load/helper.go:94
        	            				/go/testdir/integration-tests/ccip-tests/load/ccip_test.go:87
        	Error:      	Received unexpected error:
        	            	price Updates not found for token 0xDBD6c48913473F648f64d8E9fdDeead1F1734E22
        	            	waiting for price update failed on lane private-chain-5-->private-chain-19
        	            	github.com/smartcontractkit/chainlink/integration-tests/ccip-tests/testsetups.(*CCIPTestSetUpOutputs).WaitForPriceUpdates.func1
        	            		/go/testdir/integration-tests/ccip-tests/testsetups/ccip.go:916
        	            	github.com/smartcontractkit/chainlink/integration-tests/ccip-tests/testsetups.(*CCIPTestSetUpOutputs).WaitForPriceUpdates.func2
        	            		/go/testdir/integration-tests/ccip-tests/testsetups/ccip.go:922
        	            	golang.org/x/sync/errgroup.(*Group).Go.func1
        	            		/go/pkg/mod/golang.org/x/sync@v0.8.0/errgroup/errgroup.go:78
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1695
        	Test:       	TestLoadCCIPStableRequestTriggeringWithNetworkChaos
```
I've looked over the changes I've made since it was working and I'm pretty sure it doesn't change the functionality (aside from returning a sql error if there is one instead of nil, to fix an issue @dhaidashenko pointed out). Maybe another pair of eyes could help?

reductionista committed:
Fix setting of countBasedPruningActive flag

reductionista committed:
Change sql comments to go comments

mateusz-sekara commented on core/chains/evm/logpoller/log_poller.go:
```diff
@@ -545,18 +556,37 @@ func (lp *logPoller) GetReplayFromBlock(ctx context.Context, requested int64) (i
+	}
+	for _, filter := range filters {
+		if filter.MaxLogsKept != 0 {
+			lp.countBasedLogPruningActive.Store(true)
```
nit, but you can return immediately when setting that to true, right? Don't need to keep iterating

reductionista replied:
Good point!

mateusz-sekara commented on core/chains/evm/logpoller/orm.go:
```diff
@@ -320,27 +333,46 @@ func (o *DSORM) DeleteBlocksBefore(ctx context.Context, end int64, limit int64)
-		deleted += rows
-
-		if limitBlock == end {
+		if upper == end {
```
nit, but probably `if upper >= end { break }` would be safer. Not sure if there is a real threat of accidentally going beyond, but `upper >= end` will prevent from that

reductionista replied:
Just a few lines before this is:
```
		if upper > end {
			upper = end
		}
```
so definitely not possible ;-) But sure... marginally safer in case someone tries to modify the code in the future without thinking carefully

mateusz-sekara commented on core/chains/evm/logpoller/orm.go:
```diff
@@ -386,23 +418,86 @@ type Exp struct {
+		if err2 != nil {
+			return 0, err2
+		}
+		r.AddResults(rowIDs)
```
I wonder if returning `rowIDs` from `query` function wouldn't be cleaner in terms of understanding the execution flow. I don't have strong opinion on that. 

I mean this
```go
		err2 := r.ds.SelectContext(ctx, &rowIDs, query, r.chainID, lower, upper)
		if err2 != nil {
			return 0, err2
		}
		return rowIDs, int64(len(rowIDs)), nil
```

reductionista replied:
I would agree if they were all SELECT queries, but I don't like having to return an extra empty list for DELETE, INSERT, or UPDATE queries, since the # of rows affected is the only actual result in that case. Requiring the SELECT queries to call a separate method makes more sense to me, since it's an optional feature not needed for most types of queries (although we're currently only using it for DELETE and SELECT).

reductionista committed:
Address remaining PR comments

- Remove topics from SelectExcessLogs query
- Early exit from loadFilters
- upper >= end

mateusz-sekara reviewed (dismissed):
Good job. I know you also did an amazing job with playing with query analyzer and running load tests for different scenarios. It would be useful to document some of these experiments as part of this PR's description (what was tested, what are the results etc). Whoever will work on some extensions or improvements in the future will have a broader context. Also it shows some additional context to changes made ;)

dhaidashenko commented:
> As far as I can tell from the unit and integration tests, everything is passing. But since Thurs I've been having trouble getting the ccip load tests to run, so I still haven't been able to test the performance of the query when MaxLogsKept=0 for all filters. At first I was getting this error:
> 
> ```
>     log.go:43: 00:41:51.16 INF UsdPerTokenUpdated event received Lane=private-chain-10-->private-chain-8 chain=private-chain-10 env=load-ccip-3c00f price_registry=0x5E388db7c3aef7deA9585FbE3c077f0c98857d0d token=0xeD1DB453C3156Ff3155a97AD217b3087D5Dc5f6E
>     log.go:43: 00:41:51.16 INF UsdPerTokenUpdated event received Lane=private-chain-10-->private-chain-17 chain=private-chain-10 env=load-ccip-3c00f price_registry=0x5E388db7c3aef7deA9585FbE3c077f0c98857d0d token=0xeD1DB453C3156Ff3155a97AD217b3087D5Dc5f6E
> 2024/10/08 00:41:51.289506 WARN RESTY Post "http:/XX.XX.XX.XX:6688/v2/jobs": dial tcp XX.XX.XX.XX:6688: connect: connection refused, Attempt 4
> 2024/10/08 00:41:51.289534 ERROR RESTY Post "http://XX.XX.XX.XX:6688/v2/jobs": dial tcp XX.XX.XX.XX:6688: connect: connection refused
>     ccip.go:1190:
>         	Error Trace:	/go/testdir/integration-tests/ccip-tests/testsetups/ccip.go:1190
>         	            				/go/testdir/integration-tests/ccip-tests/load/helper.go:94
>         	            				/go/testdir/integration-tests/ccip-tests/load/ccip_test.go:87
>         	Error:      	Received unexpected error:
>         	            	shouldn't fail creating CCIP-Commit job on OCR node 6 job name ccip-commit-private-chain-6-private-chain-12 - Post "http://XX.XX.XX.XX:6688/v2/jobs": dial tcp XX.XX.XX.XX:6688: connect: connection refused
>         	Test:       	TestLoadCCIPStableRequestTriggeringWithNetworkChaos
>         	Messages:   	Creating jobs shouldn't fail
> ```
> 
> But after rebasing everything, I'm getting this error now (about 25 mins after the remote runner starts):
> 
> ```
>         	Error Trace:	/go/testdir/integration-tests/ccip-tests/testsetups/ccip.go:931
>         	            				/go/testdir/integration-tests/ccip-tests/testsetups/ccip.go:1192
>         	            				/go/testdir/integration-tests/ccip-tests/load/helper.go:94
>         	            				/go/testdir/integration-tests/ccip-tests/load/ccip_test.go:87
>         	Error:      	Received unexpected error:
>         	            	price Updates not found for token 0xDBD6c48913473F648f64d8E9fdDeead1F1734E22
>         	            	waiting for price update failed on lane private-chain-5-->private-chain-19
>         	            	github.com/smartcontractkit/chainlink/integration-tests/ccip-tests/testsetups.(*CCIPTestSetUpOutputs).WaitForPriceUpdates.func1
>         	            		/go/testdir/integration-tests/ccip-tests/testsetups/ccip.go:916
>         	            	github.com/smartcontractkit/chainlink/integration-tests/ccip-tests/testsetups.(*CCIPTestSetUpOutputs).WaitForPriceUpdates.func2
>         	            		/go/testdir/integration-tests/ccip-tests/testsetups/ccip.go:922
>         	            	golang.org/x/sync/errgroup.(*Group).Go.func1
>         	            		/go/pkg/mod/golang.org/x/sync@v0.8.0/errgroup/errgroup.go:78
>         	            	runtime.goexit
>         	            		/usr/local/go/src/runtime/asm_amd64.s:1695
>         	Test:       	TestLoadCCIPStableRequestTriggeringWithNetworkChaos
> ```
> 
> I've looked over the changes I've made since it was working and I'm pretty sure it doesn't change the functionality (aside from returning a sql error if there is one instead of nil, to fix an issue @dhaidashenko pointed out). Maybe another pair of eyes could help?

The changes look good to me. I do not see anything that could have caused the issue. 
I've seen the `waiting for price update failed on lane...` issue before in two cases:
1. Due to deep finality LogPoller was skipping blocks close to the genesis block. The issue was addressed, and even without the fix, it should not occur on normal finality.
2. On restart of Anvil pod (not sure what causes it), Anvil was dropping blocks history. This caused LP to halt with `Reorg greater than finality depth`. The issue was fixed with an update of the Anvil chart, but maybe it was reintroduced in a newer version, or we've rolled back to the old version.

jmank88 commented on core/chains/evm/logpoller/log_poller.go:
```diff
@@ -134,7 +134,8 @@ type logPoller struct {
 	// LogPoller keeps running in infinite loop, so whenever the invalid state is removed from the database it should
 	// recover automatically without needing to restart the LogPoller.
-	finalityViolated *atomic.Bool
+	finalityViolated           *atomic.Bool
+	countBasedLogPruningActive *atomic.Bool
```
nit/ These could be values, then the zero value will work without requiring initialization in the constructor.
```suggestion
	finalityViolated           atomic.Bool
	countBasedLogPruningActive atomic.Bool
```

reductionista committed:
Take Jordan's suggestion of replacing *atomic.Bool with atomic.Bool
